### PR TITLE
[docs] Improved the 'Recipes/React' documentation

### PR DESCRIPTION
**Pull request by mattpocock**

### CODE REVIEW

File: docs/recipes/react.md [hunk 1]
```diff
@@ -1,31 +1,20 @@
 # Usage with React
 
-The most straightforward way of using XState with React is through local component state. The machine used should always be decoupled from implementation details; e.g., it should never know that it is in React (or Vue, or Angular, etc.):
+XState can be used with React to:
 
-```js
-import { createMachine } from 'xstate';
+- Co-ordinate local state
```
Comment (davidkpiano) [inline docs/recipes/react.md:5]:
```suggestion
- Coordinate local state
```

American spelling? 😜

Comment (mattpocock) [reply docs/recipes/react.md:5]: 😂  nope, just wrong spelling

Had my grammar corrected by an American - bad form, Matthew

File: docs/recipes/react.md [hunk 2]
```diff
@@ -1,31 +1,20 @@
...
-    },
-    active: {
-      on: { TOGGLE: 'inactive' }
-    }
-  }
-});
-```
+At [Stately](https://stately.ai), we love this combo. It's our go-to stack for creating internal applications.
+
+To ask for help, check out the `#react-help` channel on [Discord](https://discord.com/invite/xstate).
```
Comment (davidkpiano) [inline docs/recipes/react.md:11]:
```suggestion
To ask for help, check out the [`#react-help` channel in our Discord community](https://discord.gg/vedXj62MfQ).
```

Channel-specific invite link (never expires)

File: docs/recipes/react.md [hunk 3]
```diff
@@ -40,8 +29,186 @@ function Toggle() {
...
+    <GlobalStateContext.Provider value={{ authService }}>
+      {props.children}
+    </GlobalStateContext.Provider>
+  );
+};
+```
+
+Using `useInterpret` returns a service, which is a static reference to the running machine which can be subscribed to. This value never changes, so we don't need to worry about wasted re-renders.
+
+### Utilising context
```
Comment (davidkpiano) [inline docs/recipes/react.md:64]:
```suggestion
### Utilizing context
```

File: docs/recipes/react.md [hunk 4]
```diff
@@ -40,8 +29,186 @@ function Toggle() {
...
+There's an issue with the implementation above - this will update the component for any change to the service. Tools like [Redux](https://redux.js.org) use [`selectors`](https://redux.js.org/usage/deriving-data-selectors) for deriving state. Selectors are functions which restrict which parts of the state can result in components re-rendering.
+
+Fortunately, XState exposes the `useSelector` hook.
+
+```js
+import React, { useContext } from 'react';
+import { GlobalStateContext } from './globalState';
+import { useSelector } from '@xstate/react';
+
+const selector = (state) => {
```
Comment (davidkpiano) [inline docs/recipes/react.md:94]:
```suggestion
const loggedInSelector = (state) => {
```

File: docs/recipes/react.md [hunk 5]
```diff
@@ -40,8 +29,186 @@ function Toggle() {
...
+  return state.matches('loggedIn');
+};
+
+export const SomeComponent = (props) => {
+  const globalServices = useContext(GlobalStateContext);
+  const isLoggedIn = useSelector(globalServices.authService, selector);
```
Comment (davidkpiano) [inline docs/recipes/react.md:100]:
```suggestion
  const isLoggedIn = useSelector(globalServices.authService, loggedInSelector);
```

**Review by davidkpiano — APPROVED** [review]

**Comment by davidkpiano** [comment]
Really good improvement over the current docs; we can always tweak later 👍

**Pull request merged**